lib/Optimizer/GraphOptimizer.cpp

```diff
@@ -2512,6 +2445,87 @@ static void foldLeakyRelu(Function *F) {
   }
 }
 
+/// Parameters that are used to define ChannelShuffle operators.
+struct ChannelShuffleParams {
+  size_t group;
+  size_t kernel;
+};
+
+/// Compute the original parameters to the ChannelShuffle operator (represented
+/// as ReshapeNode->TransposeNode->ReshapeNode) for which \p node is the leading
+/// ReshapeNode. \returns The original ChannelShuffle parameters if possible and
+/// empty Optional otherwise.
+static llvm::Optional<ChannelShuffleParams>
+getChannelShuffleParams(const ReshapeNode &node) {
+  auto resM = llvm::Optional<ChannelShuffleParams>();
+
+  llvm::ArrayRef<size_t> inputDims = node.getInput().dims();
+  llvm::ArrayRef<size_t> resultDims = node.getDims();
+
+  // Check that there is one more output dimension than input dimension.
+  if (resultDims.size() != inputDims.size() + 1) {
+    return resM;
+  }
+
+  // Find the first output dimension that doesn't match its corresponding input
+  // dimension.
+  ChannelShuffleParams params;
+  bool found = false;
+  for (size_t i = 0, e = resultDims.size(); i < e - 1; ++i) {
+    if (inputDims[i] != resultDims[i]) {
+      params.kernel = i;
+      params.group = resultDims[i];
+      found = true;
+      break;
+    }
+  }
+
+  // Double check the property that the mismatched output found dimension and
+  // its successor together evenly multiply to the input dimension they
+  // mismatched on.
+  if (found && resultDims[params.kernel] * resultDims[params.kernel + 1] ==
+                   inputDims[params.kernel]) {
+    resM = params;
+  }
+
+  return resM;
+}
+
+// Fold Reshape->Transpose->Reshape into ChannelShuffle when applicable.
+static void foldChannelShuffle(Function *F) {
+
+  auto &nodes = F->getNodes();
+  for (auto &node : nodes) {
+    auto *RN2 = dyn_cast<ReshapeNode>(&node);
+    if (!RN2 || RN2->getDims().size() != 4) {
```

Comment:
I don't think it's a requirement that the ChannelShuffle input has 4 dimensions, is it? I think we can drop the checks for the dims size and just let `getChannelShuffleParams()` handle the logic for checking if it's a valid ChannelShuffle.